homeassistant/components/edl21/sensor.py

```diff
@@ -0,0 +1,164 @@
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.typing import Optional
+from homeassistant.util import Throttle
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = "edl21"
+CONF_SERIAL_PORT = "serial_port"
+ICON_POWER = "mdi:flash"
+MIN_TIME_BETWEEN_UPDATES = timedelta(seconds=60)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({vol.Required(CONF_SERIAL_PORT): cv.string})
+
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Set up the EDL21 sensor."""
+    hass.data[DOMAIN] = EDL21(hass, config, async_add_entities)
+    await hass.data[DOMAIN].connect()
+
+
+class EDL21:
+    """EDL21 handles telegrams sent by a compatible smart meter."""
+
+    # OBIS format: A-B:C.D.E*F
+    _OBIS_NAMES = {
+        # A=1: Electricity
+        # C=0: General purpose objects
+        "1-0:0.0.9*255": "Electricity ID",
+        # C=1: Active power +
+        # D=8: Time integral 1
+        # E=0: Total
+        "1-0:1.8.0*255": "Positive active energy total",
+        # E=1: Rate 1
+        "1-0:1.8.1*255": "Positive active energy in tariff T1",
+        # E=2: Rate 2
+        "1-0:1.8.2*255": "Positive active energy in tariff T2",
+        # D=17: Time integral 7
+        # E=0: Total
+        "1-0:1.17.0*255": "Last signed positive active energy total",
+        # C=15: Active power absolute
+        # D=7: Instantaneous value
+        # E=0: Total
+        "1-0:15.7.0*255": "Absolute active instantaneous power",
+        # C=16: Active power sum
+        # D=7: Instantaneous value
+        # E=0: Total
+        "1-0:16.7.0*255": "Sum active instantaneous power",
+    }
+
+    def __init__(self, hass, config, async_add_entities) -> None:
+        """Initialize an EDL21 object."""
+        self._blacklist = {
+            # A=129: Manufacturer specific
+            "129-129:199.130.3*255",  # Iskraemeco: Manufacturer
+            "129-129:199.130.5*255",  # Iskraemeco: Public Key
+        }
+        self._cache = {}
+        self._hass = hass
+        self._async_add_entities = async_add_entities
+        self._proto = SmlProtocol(config[CONF_SERIAL_PORT])
+        self._proto.add_listener(self.event, ["SmlGetListResponse"])
+
+    async def connect(self):
+        """Connect to an EDL21 reader."""
+        await self._proto.connect(self._hass.loop)
+
+    def event(self, message_body) -> None:
+        """Handle events from pysml."""
+        assert isinstance(message_body, SmlGetListResponse)
+
+        new_devices = []
+        for telegram in message_body.get("valList", []):
+            obis = telegram.get("objName")
+            if not obis:
+                continue
+
+            device = self._cache.get(obis)
+            if not device:
+                name = self._OBIS_NAMES.get(obis)
+                if name:
+                    device = self._cache[obis] = EDL21Entity(obis, name, telegram)
```

Comment:
If we use our dispatch helper below we don't need to store the entity instance, just store a reference to the entity eg `obis`.